### ISSUE TITLE
BMP image does not read correctly anymore.

### ISSUE THREAD

**Issue by dsa-t** [wxWidgets/wxWidgets#24219]
A specific BMP image does not read correctly anymore.

Here's how the image is supposed to look like:

![image](https://github.com/wxWidgets/wxWidgets/assets/37658952/2aafb092-1712-47ee-997d-6cc7d99f5d47)

Most colors are replaced with transparency when loaded by wx and saved as png:

![image](https://github.com/wxWidgets/wxWidgets/assets/37658952/aaf0f061-4776-4309-b38e-1d3f8577168d)

Original table.bmp and the result saved in png is in the archive:
[image.zip](https://github.com/wxWidgets/wxWidgets/files/13940674/image.zip)

### Platform and version information

- wxWidgets version  you use: 3.3 (8967bed1904a8cabe72a68e8e36155ef307a7086)
- wxWidgets port  you use: wxGTK
- OS  and its version: Linux Mint 21.1

**Comment by paulcor** [comment]
Can be seen with the `image` sample on MSW, so not GTK-specific.

**Comment by Randalphwa** [comment]
Placing the BMP file into the richtext sample just shows a black box (run on Windows 11). The BMP does display correctly when loaded into Windows 11 Paint. It does _not_ display correctly in Affinity Photo 2, Affinity Designer 2, or Topaz Sharpen AI. It does display correctly in HTML Help Image Editor which has not been updated since 1999 -- so while the BMP format might be unusual enough to throw off the commercial products I tried it on, it's a format that was at least understandable 25 years ago.

**Comment by PBfordev** [comment]
> A specific BMP image does not read correctly **anymore**

The image by the OP is on Windows displayed by the image sample in grayscale (i.e., incorrectly) from wxWidgets 3.0.5 to the (almost) current GIT master (tested also many versions in between, including 3.2.x and bunch of 3.1.x), so it is not a recent thing, at least on MSW.

**Comment by dsa-t** [comment]
> so it is not a recent thing, at least on MSW.

You're right. I was looking at wrong data.

**Comment by dsa-t** [comment]
> Does the image have alpha or not?

It doesn't seem like it's supposed to have alpha. GIMP doesn't import an alpha channel with it.

**Comment by PBfordev** [comment]
FWIW, I have tried a bunch of programs on Windows 10 and they all displayed the image in color:
FireFox and Chrome, GIMP, Corel PhotoPAINT,  Inkscape,  IrfanView, Total Commander, Paint, Paint 3D, Photos, Snip&Sketch, Libre Office and Microsoft Office.

The only program that displayed the image with transparency was web-based [Photopea](https://www.photopea.com/).

So I think, the problem is not with the image but wxWidgets. I also tried the image sample from wxWidgets 2.8.12 and it has issues to display the image at all, so...

**Comment by paulcor** [comment]
The file loads correctly on Windows using `wxBitmap::LoadFile()`, which uses WIN32 `LoadFile()`. I think the issue here is that despite the depth being 32, Windows (and most other software) only uses the RGB values and not the alpha. Maybe the alpha is only used if the file has an "extended" header, like `BITMAPV4HEADER`.

**Comment by paulcor** [comment]
Actually, I think the alpha is only used if `BITMAPINFOHEADER.biCompression == BI_BITFIELDS`.

**Comment by vadz** [comment]
This was supposed to be fixed by the commits above, right? So I'm closing it and will backport the fix to 3.2 where it would seem to be needed too. Please let me know if I'm missing anything.

**Comment by paulcor** [comment]
There were a number of later commits to finally get it right, culminating in
e7e352d5f4 (Handle BMP alpha the same way Windows does, 2025-02-22)
and the followup
4d5fbf5afa (Fix BMP test under Wine, 2025-02-22)

**Comment by vadz** [comment]
Yes, I hope I've got all of them. Please try to review my next 3.2 backports branch if you can.
